packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts

```diff
@@ -0,0 +1,325 @@
+    };
+
+    return ['udp', 'tcp'].every((protocol) => expecting[protocol] === gotResult[protocol]);
+  }
+
+  private registerIceCandidateListener() {
+    this.pc.onicecandidate = (e) => {
+      const CANDIDATE_TYPES = {
+        SERVER_REFLEXIVE: 'srflx',
+        RELAY: 'relay',
+      };
+
+      if (e.candidate) {
+        if (e.candidate.type !== 'host') {
+          console.log('marcin: e.candidate', e.candidate);
+        }
+        if (e.candidate.type === CANDIDATE_TYPES.SERVER_REFLEXIVE) {
+          if (this.result.udp.latencyInMilliseconds === undefined) {
+            const elapsed = this.getElapsedTime();
+
+            LoggerProxy.logger.log(
+              // @ts-ignore
+              `Reachability:index#onIceCandidate --> Successfully reached ${this.name} over UDP: ${elapsed}ms`
+            );
+            this.result.udp.latencyInMilliseconds = elapsed.toString();
+            this.result.udp.reachable = 'true';
+          }
+          this.addPublicIP('udp', e.candidate.address);
+        }
+
+        if (e.candidate.type === CANDIDATE_TYPES.RELAY) {
+          if (this.result.tcp.latencyInMilliseconds === undefined) {
+            const elapsed = this.getElapsedTime();
+
+            LoggerProxy.logger.log(
+              // @ts-ignore
+              `Reachability:index#onIceCandidate --> Successfully reached ${this.name} over TCP: ${elapsed}ms`
+            );
+            console.log(`marcin: ${this.name}: TCP: e.candidate`, e.candidate);
+            this.result.tcp.latencyInMilliseconds = elapsed.toString();
+            this.result.tcp.reachable = 'true';
+          }
+          this.addPublicIP('tcp', e.candidate.address);
+        }
+
+        if (this.haveWeGotAllResults()) {
+          this.closePeerConnection();
+          this.finishReachabilityCheck();
+        }
+      }
+    };
+  }
+
+  /**
+   * Starts the process of doing reachability checks on the media cluster.
+   *
+   * @returns {Promise}
+   */
+  async start(): Promise<ReachabilityResult> {
+    if (!this.pc) {
+      LoggerProxy.logger.warn(
+        `Reachability:ClusterReachability#start --> Error: peerConnection is undefined`
+      );
+
+      return this.result;
+    }
+
+    // Initialize this.result as saying that nothing is reachable.
+    // It will get updated as we go along and successfully gather ICE candidates.
+    this.result = {
+      udp: {
+        reachable: this.numUdpUrls > 0 ? 'false' : undefined,
+        untested: this.numUdpUrls === 0 ? 'true' : undefined,
+      },
+      tcp: {
+        reachable: this.numTcpUrls > 0 ? 'false' : undefined,
+        untested: this.numTcpUrls === 0 ? 'true' : undefined,
+      },
+      xtls: {
+        untested: 'true',
+      },
+    };
+
+    try {
+      const offer = await this.pc.createOffer({offerToReceiveAudio: true});
+
+      this.startTimestamp = performance.now();
+
+      this.pc.setLocalDescription(offer); // not awaiting on purpose
```

Comment:
what is the purpose

Reply:
the call to `setLocalDescription()` with the offer triggers browser to start gathering ICE candidates and contacting STUN and TURN servers

Reply:
we don't await it, because we don't care about it completing, we're just waiting for the ICE candidates that we get through the listeners

Reply:
I think the comment here explaining it would be useful

Reply:
done